core/src/apps/bitcoin/authorization.py

```diff
@@ -0,0 +1,61 @@
+from micropython import const
+
+from trezor.messages import MessageType
+from trezor.messages.TxInputType import TxInputType
+
+from .common import BIP32_WALLET_DEPTH
+
+if False:
+    from typing import Iterable
+    from trezor.messages.AuthorizeCoinJoin import AuthorizeCoinJoin
+    from trezor.messages.GetOwnershipProof import GetOwnershipProof
+    from trezor.messages.SignTx import SignTx
+    from apps.common import coininfo
+    from apps.common.seed import Keychain
+
+_ROUND_ID_LEN = const(32)
+FEE_PER_ANONYMITY_DECIMALS = const(9)
+
+
+class CoinJoinAuthorization:
+    def __init__(
+        self, msg: AuthorizeCoinJoin, keychain: Keychain, coin: coininfo.CoinInfo
+    ):
+        self.coordinator = msg.coordinator
+        self.remaining_fee = msg.max_total_fee
+        self.fee_per_anonymity = msg.fee_per_anonymity or 0
+        self.address_n = msg.address_n
+        self.keychain = keychain
+        self.coin = coin
+        self.script_type = msg.script_type
+
+    def __del__(self) -> None:
+        self.keychain.__del__()
+
+    def expected_wire_types(self) -> Iterable[int]:
+        return (MessageType.SignTx, MessageType.GetOwnershipProof)
+
+    def check_get_ownership_proof(self, msg: GetOwnershipProof) -> bool:
+        # Check whether the current authorization matches the parameters of the request.
+        return (
+            msg.address_n[:-BIP32_WALLET_DEPTH] == self.address_n
```

Comment:
this and below should probably also check that `len(msg.address_n) > BIP32_WALLET_DEPTH`

Reply:
perhaps even `len(msg.address_n) == len(self.address_n) + BIP32_WALLET_DEPTH` ?

Reply:
Good point. As for the two options:
- `len(msg.address_n) >= BIP32_WALLET_DEPTH`
- `len(msg.address_n) == len(self.address_n) + BIP32_WALLET_DEPTH`

These are logically equivalent given that we check `msg.address_n[:-BIP32_WALLET_DEPTH] == self.address_n`. In case of the first option you actually used `>` instead of `>=`, which would ensure that the length of `self.address_n` is at least 1. Is that intentional? It makes sense to put in place at least some basic restriction like this one, but then it should go into `authorize_coinjoin()`.

Reply:
that's intentional, and yes, that should be part of authorize_coinjoin.

I think I like the second option better because it's more explicit about the intended state

Reply:
https://github.com/trezor/trezor-firmware/pull/1127/commits/67e85197e3811ff3bafe000a5c07ae775ca4f62a and https://github.com/trezor/trezor-firmware/pull/1127/commits/e13abfc57b4b95cffb76155489ba105783a057d3, I already used the first option